Atomate2 OpenMM integration & broader classical MD framework

orionarcher opened the pull request:
Developed by @xperrylinn and @orionarcher

## Summary

This PR builds out support for OpenMM with a framework that could be extended to support other MD codes. Namely LAMMPS, Amber, and Gromacs. Rough visual example [here](https://www.figma.com/file/aIwwYaPOeRPytAVNwLewOS/MD-Workflow-Diagram?type=whiteboard&node-id=0%3A1&t=rXTgEcKkXIyKnzes-1)

### Core ideas:
 - Use the  `openff.interchange.Interchange` object as a core engine-agnostic representation of an MD simulation.
 - Define a generic `ClassicalMDTaskDocument` and generic `Interchange` creation functions, then hand off responsibility for evolving the simulation to workflows for each MD engine. Currently only OpenMM support is implemented.
 - Support low-throughput workflows by making workflows that can easily output to local directories.
 - Use `prev_task` to pass meta-data between jobs. Since OpenMM is programmatic, the core information is not stored in the local directory.
 - No `InputSets`. Since OpenMM is programmatic, `InputSets` and `InputGenerators` don't make conceptual sense. 

### What's implemented:
- Integration of Atomate2 and OpenFF for classical MD simulations
- Support for energy minimization, NPT, NVT, and temperature change jobs
- Implementation of annealing and production workflows
- Serialization and deserialization of OpenFF objects (Molecule, Topology, Interchange, Quantity) with `monty`
- Utility functions for generating OpenFF Interchange objects and creating molecule specifications

### To do / open questions:
- [x] Need to add create conditional imports that fail nicely, where are some examples of this?


### Future PR
- [x] Store trajectories in `additional_stores`, if available
- [x] The core `openff.interchange.Interchange` object can be larger than 16MB MongoDB doc limit. Need to implement a workaround by zipping the object and/or putting it on S3.
- [x] migrate schema dependencies to emmet
- [x] migrate utility dependencies to pymatgen


### Related PRs:
- PR to `pymatgen` would move most of `atomate2.classical_md.utils` upstream. [PR #3729](https://github.com/materialsproject/pymatgen/pull/3729)
- PR to `emmet` would move `atomate2.classical_md.schemas` upstream. [PR #975](https://github.com/materialsproject/emmet/pull/975)

### Example usage


```python
mol_specs_dicts = [
    {"smile": "CCO", "count": 10, "name": "ethanol"},
    {"smile": "O", "count": 200, "name": "water"},
]
inter_job = generate_interchange(mol_specs_dicts, mass_density=1)

production_maker = ProductionMaker(
    energy_maker=EnergyMinimizationMaker(),
    npt_maker=NPTMaker(steps=300000, pressure=1),
    anneal_maker=AnnealMaker.from_temps_and_steps(
        steps=1000000, anneal_temp=400, final_temp=300
    ),
    nvt_maker=NVTMaker(steps=5000000),
)

production_flow = production_maker.make(
    inter_job.output.interchange, 
    output_dir="my/directory/"
    prev_task=inter_job.output
)

wf = Flow([inter_job, production_flow])

run_locally(wf)

# will put outputs for the whole flow in "my/directory/"
```


## Additional dependencies introduced

* openmm
* openff-toolkit
* openff-interchange
* packmol
* openbabel

These are all necessary for the classical MD setup and execution workflow.

## Checklist

Before a pull request can be merged, the following items must be checked:

* [x] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/).
  The easiest way to handle this is to run the following in the **correct sequence** on
  your local machine. Start with running [`ruff`](https://docs.astral.sh/ruff) and `ruff format` on your new code. This will
  automatically reformat your code to PEP8 conventions and fix many linting issues.
* [x] Doc strings have been added in the [Numpy docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html).
  Run [ruff](https://beta.ruff.rs/docs/rules/#pydocstyle-d) on your code.
* [x] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org) to
  type check your code.
* [x] Tests have been added for any new functionality or bug fixes.
* [x] All linting and tests pass.

Note that the CI system will run all the above checks. But it will be much more
efficient if you already fix most errors prior to submitting the PR. It is highly
recommended that you use the pre-commit hook provided in the repository. Simply run
`pre-commit install` and a check will be run prior to allowing commits.

orionarcher commented:
Hey @utf, this PR supercedes #717 and is now ready for feedback. Testing is passing locally but I need to fix some issues with the CI. I also need to flesh out the documentation/tutorials.

@rkingsbury might be of interest to you as well!

orionarcher commented:
The OpenForceField toolkit can only be installed via conda, making the current testing framework unsuitable. 

I initially tried migrating the whole test suite to use `micromamba` instead of pip. I got almost all of the tests passing but could not debug the tests for `forcefields`. Since I wasn't sure you'd want to migrate even if it worked, I decided to drop that line of work. Instead, I decided to exclude the `classical_md` tests from the `test` action and create a new action called `test_classical_md` that uses `micromamba` to install dependencies and then only executes the `classical_md` tests. This seems to be working.

orionarcher commented:
I am currently working on a builder for the classical_md code. I plan to stage the builder in `atomate2` for now. Long term, I plan to migrate most of the code in `utils.py` to `pymatgen` and the builder code and TaskDoc code to `emmet`, mirroring the organization for the rest of the MP codes.

orionarcher commented:
As far as I can tell the failing tests are upstream and not due to this PR.

rkingsbury commented:
Looks like a great contribution @orionarcher . I'm especially glad you included a tutorial as part of the PR. Is there any aspect you or @utf would find helpful for me to review specifically?

orionarcher commented:
From @rkingsbury, I'd love it if you could take a look at the pydantic schemas (now moved to emmet, [here](https://github.com/materialsproject/emmet/pull/975/files)). Your MD experience would add a useful perspective.

@utf, it'd be very helpful to get your feedback on the changes to `testing.yml`. I added a new action and want to make sure it looks good to you.

Overall, I think this is pretty much done and could be merged. The outstanding features aren't critical and could be added in a later PR.

rkingsbury commented:
> From @rkingsbury, I'd love it if you could take a look at the pydantic schemas (now moved to emmet, [here](https://github.com/materialsproject/emmet/pull/975/files)). Your MD experience would add a useful perspective.

Done! I left a review on the already merged PR here - https://github.com/materialsproject/emmet/pull/975#pullrequestreview-1994562429

orionarcher commented:
@utf would you mind taking a look at this?

I broke compatibility between emmet and atomate2 in the most recent commits but I will resolve it soon. If you'd prefer to look at something that is passing, please consider [commit d0683c9](https://github.com/materialsproject/atomate2/pull/782/commits/d0689c36a9c047e61d906b47ee8db0bc6c092d16) which has all of the core functionality.

utf commented on docs/user/codes/openmm.md:
```diff
@@ -0,0 +1,639 @@
+from emmet.builders.classical_md.utils import create_universe, create_solute
+from openff.interchange import Interchange
+
+ec_emc_taskdoc = ClassicalMDTaskDocument.parse_file("tutorial_system/taskdoc.json")
```
Better to just get the task doc from the database rather than relying on folders.

utf commented on pyproject.toml:
```diff
@@ -57,6 +57,7 @@ forcefields = [
     "quippy-ase>=0.9.14",
     "torch<=2.2.1",       # incompatibility with dgl if newer versions are used
 ]
+classical_md = ["openmm-mdanalysis-reporter>=0.1.0"]
```
Put all other requirements here.

utf replied:
Add other requirements here.

utf commented on src/atomate2/classical_md/__init__.py:
```diff
@@ -0,0 +1,70 @@
+    return mol_dict
+
+
+Molecule.as_dict = openff_mol_as_monty_dict
```
I'm not sure how I feel about this. Seems like it could break easily? The other options are: 
1. Add as_dict methods to openff classes for monty support.
2. Add support for openff to monty (doesn't seem likely given we tried to do something similar for ase and it was turned down).
3. Some kind of plugin system for monty where we can define some common serialisation routes for materials science objects.

For now we can leave this as is but I'd like to explore option 3 further (cc'ing @Andrew-S-Rosen @janosh @mkhorton)

orionarcher replied:
Agreed that it's not a perfect solution.

I think I'd be turned down for 1 & 2 both. I'd be happy to support option 3 in the future.

utf commented on src/atomate2/classical_md/__init__.py:
```diff
@@ -0,0 +1,70 @@
+"""Module for classical md workflows."""
```
To me this folder should not be called `classical_md` but instead `openff`. And the `openmm` subdirectory should be its own folder in the base dir. E.g., `atomate2/openmm`

utf commented on src/atomate2/classical_md/openmm/flows/core.py:
```diff
@@ -0,0 +1,159 @@
+
+
+@dataclass
+class OpenMMFlowMaker:
```
This is an interesting approach to generate a flow but makes sense in the context!

utf commented on src/atomate2/classical_md/openmm/flows/core.py:
```diff
@@ -0,0 +1,159 @@
+        self,
+        interchange: Interchange | bytes,
+        prev_task: ClassicalMDTaskDocument | None = None,
+        output_dir: str | Path | None = None,
```
Can we remove output_dir from this. We should handle this another way.

utf commented on src/atomate2/classical_md/openmm/flows/core.py:
```diff
@@ -0,0 +1,159 @@
+        )
+
+    @classmethod
+    def anneal_flow(
```
Would it make sense to turn this into's its own class?

orionarcher replied:
Let me think about this. I agree that a monolithic flow maker isn't great design.

utf commented on src/atomate2/classical_md/openmm/jobs/base.py:
```diff
@@ -0,0 +1,494 @@
+if TYPE_CHECKING:
+    from openmm.app.simulation import Simulation
+
+OPENMM_MAKER_DEFAULTS = {
```
What is the benefit of having these here rather than settings them in the `BaseOpenMMMaker` directly? It seems like this would be cleaner?

orionarcher replied:
I agonized over this but it was ultimately the only solution I could think of. I'll respond here but the answer would be the same for this comment too:

> I think this would be even simpler if the defaults are on the base class, as then there are only two places to check.

When users are defining workflows I wanted to support easy inheritance between jobs. Priority is: settings from current maker > settings from previous task input >  a default value. If I set the defaults in BaseOpenMMMaker directly, there is no way of differentiating between settings from the current maker and the defaults. Maybe there is another solution, but the only one I could think of was external defaults. 

You can see the inheritance logic below.

```python
        if getattr(self, attr, None) is not None:
            attr_value = getattr(self, attr)  # if the defaults were set directly this would always return them.
        elif getattr(prev_input, attr, None) is not None:
            attr_value = getattr(prev_input, attr)
        else:
            attr_value = defaults.get(attr)
```

utf commented on src/atomate2/classical_md/openmm/jobs/base.py:
```diff
@@ -0,0 +1,494 @@
+    """
+
+    name: str = "base openmm job"
+    tags: list[str] | None = field(default=None)
```
I.e., the `field` could be replaced with the default.

utf commented on src/atomate2/classical_md/openmm/jobs/base.py:
```diff
@@ -0,0 +1,494 @@
+        else:
+            interchange = copy.deepcopy(interchange)
+
+        dir_name = Path(
```
I wouldn't expect to create a new directory? This will already be handled by the workflow manager (e.g., jobflow, jobflow-remote, or fireworks). I think we should remove this to be consistent with the other workflows.

utf commented on src/atomate2/classical_md/openmm/jobs/base.py:
```diff
@@ -0,0 +1,494 @@
+        task_doc = self._create_task_doc(interchange, elapsed_time, dir_name, prev_task)
+
+        # write out task_doc json to output dir
+        with open(dir_name / "taskdoc.json", "w") as file:
```
Why do we write the task doc? I'm not entirely opposed to this but it will get stored in the database anyway.

orionarcher replied:
Mainly to enable easier local analysis.

utf commented on src/atomate2/classical_md/openmm/jobs/base.py:
```diff
@@ -0,0 +1,494 @@
+        prev_task: OpenMMTaskDocument | None = None,
+        add_defaults: dict | None = None,
+    ) -> Any:
+        """Resolve an attribute and set its value.
```
I think this would be even simpler if the defaults are on the base class, as then there are only two places to check.

orionarcher replied:
See my response above RE:

> What is the benefit of having these here rather than settings them in the BaseOpenMMMaker directly? It seems like this would be cleaner?

utf commented on tutorials/openmm_tutorial.ipynb:
```diff
@@ -0,0 +1,605 @@
+{
```
This is awesome. But I think this should live in the docs folder somewhere. E.g., for jobflow we have the `jobflow/docs/tutorials` folder that contains python notebooks, and these are automatically built into the docs.

orionarcher replied:
Do you have a particular preference? I chose the location based on @jmmshn's tutorial PR but can definitely put it elsewhere. 

I would love to see automatically built jupyter notebooks for tutorials and would happily rewrite the docs into a notebook if it was available.

utf reviewed (commented):
Hi @orionarcher, appologies it has taken quite some time to get to this PR.

Firstly, it looks fantastic. This is a serious amount of work, so thank you for contributing it to atomate2. I particularly like all the documentation you have added that covers the setup, running and even builders.

My comments are mainly related to bringing some quirks of these workflows more inline with the rest of atomate2. These are primarily:
- Updating the requirements/tests so that there isn't a separate mode of installing/testing the md workflows.
- Removing the folder control from the workflows. This is something that should be handled by the workflow manager. I'd strongly encourage you to use the database to get the outputs in your examples, rather than using the file system.
- Renaming classical_md to openff, and making openmm it's own folder in the base directory.

I'm happy to chat futher about any of these issues if you want to call. Once these changes are done (and all tests pass) I'd be happy to merge.

orionarcher commented on .github/classical_md_requirements.txt:
```diff
@@ -0,0 +1,8 @@
+# for consistency, should align with pyproject.toml
+openmm>=8.1.0
```
The problem is that all of the dependencies in this file are not pip installable. They are only available on conda.

orionarcher commented on .github/workflows/testing.yml:
```diff
@@ -74,7 +74,39 @@ jobs:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          file: ./coverage.xml
+
+  test_md:
```
Many of the classical_md dependencies are not pip installable. We could consolidate the tests, but I'd need to modify the testing action to use micromamba instead of pip. I'll say more about this in another comment below.

orionarcher commented on docs/user/codes/openmm.md:
```diff
@@ -0,0 +1,639 @@
+>>> conda activate atomate2
+
+# installing atomate2
+>>> pip install git+https://github.com/orionarcher/atomate2.git
```
Will update.

orionarcher commented on docs/user/codes/openmm.md:
```diff
@@ -0,0 +1,639 @@
+from atomate2.classical_md.core import generate_interchange
+
+mol_specs_dicts = [
+    {"smile": "O", "count": 200, "name": "water"},
```
Good catch. Will update.

orionarcher commented on docs/user/codes/openmm.md:
```diff
@@ -0,0 +1,639 @@
+production_flow = production_maker.make(
+    elyte_interchange_job.output.interchange,
+    prev_task=elyte_interchange_job.output,
+    output_dir="./tutorial_system",
```
Might it still make sense to store this information?

I am considering removing the argument and then interally replacing:

```python
dir_name = Path(
    output_dir or getattr(prev_task, "dir_name", None) or Path.cwd()
)
```

with

```python
dir_name = Path.cwd()
```

utf replied:
It still makes sense to put this in the task document but not to have it controllable by the job. You can see how we did this in the FHI-aims task document: https://github.com/materialsproject/atomate2/blob/e210b58b93d33b13af800dd001c9689f96a6eb0c/src/atomate2/aims/schemas/task.py#L326

orionarcher commented:
Hi @utf, I've addressed several of the points you mentioned and am still mulling over the best solution for others. In the meantime, I'd appreciate your help with two issues.

1. I migrated the testing to use micromamba and uv. This cuts the "Install dependencies" step from ~150 seconds to ~50 seconds. Further, the migration only broke a single [cclib-related-test](https://github.com/materialsproject/atomate2/actions/runs/10117212365/job/27981713046). If you could fix it or tag the appropriate code owner, that'd be much appreciated. 

For now, I am ignoring the `classical_md` tests because they are dependent on upstream changes in `emmet` that have not been merged in.

2. Currently, `calcs_reversed` is being used to store all calculations that preceded the task, which is not the intended behavior. I want to change that while preserving the job's history, ideally by storing all previous `job_uuid`. Is that possible? Can I access the `job_uuid` from inside the maker and store it in the `TaskDoc`? If not, how should I link the job to previous jobs?

orionarcher commented:
I've now addressed all the points you brought up in your review. Most significantly:
- splitting `classical_md` into `openmm` and `openff`.
- replacing history tracking using calcs_reversed with a final `collect_jobs` job that gathers uuids and calcs
- reworking the CI and dependencies

Once emmet is adjusted and tests are passing, I think this will be ready to merge.

utf commented on src/atomate2/openmm/jobs/base.py:
```diff
@@ -0,0 +1,528 @@
+        state_file_name = self._resolve_attr("state_file_name", prev_task)
+        traj_file_name = self._resolve_attr("traj_file_name", prev_task)
+        traj_file_type = self._resolve_attr("traj_file_type", prev_task)
+        calc = Calculation(
```
Ideally the sub task documents should be created in a function in `OpenMMTaskDocument`. E.g., `OpenMMTaskDocument.from_directories()`. Can you move this logic there? The role of the `_create_task_doc` would then be to just resolve the previous directory names etc.

orionarcher replied:
I considered doing this originally but I think it'd be a headache. We could only reconstitute the information purely from the directories if we already had the current `task_doc` which is what we need to write. I'm sure there is some hack around but ultimately I feel this is where the logic belongs.

OpenMM has no input sets so the input logic is never fully encoded as a string. To get the task doc from a directory, we'd almost need to already have the taskdoc in the directory.

utf commented on src/atomate2/openmm/jobs/opls.py:
```diff
@@ -0,0 +1,256 @@
+"""Utilities for working with the OPLS forcefield in OpenMM."""
```
Currently this file is in the `jobs` folder but nothing here is a `job`. Perhaps this should be one directory up and called `utils.py`?

orionarcher replied:
`generate_openmm_interchange` is a job but I do agree some of this could be moved upstream to utils.

utf reviewed (commented):
Thanks @orionarcher, I'll get back to you about the failing test separately.

utf commented on .github/workflows/testing.yml:
```diff
@@ -31,58 +31,97 @@ jobs:
         if: github.event_name == 'repository_dispatch' && github.event.action == 'pymatgen-ci-trigger'
-        run: pip install --upgrade 'git+https://github.com/materialsproject/pymatgen@${{ github.event.client_payload.pymatgen_ref }}'
+        run: |
+          micromamba activate a2
```
Do we actually need all of these `micromamba active` commands. I think the first one should be enough.

orionarcher replied:
I tried removing them and ran into errors

janosh replied:
confirm you have to activate envs in every step, there's no propagation of shell state between job steps in GH Actions

utf commented on .github/workflows/testing.yml:
```diff
@@ -31,58 +31,97 @@ jobs:
-          pip install .[strict,tests,abinit]
-          pip install torch-runstats
-          pip install --no-deps nequip==0.5.6
+          uv pip install .[strict,tests,abinit,classical_md]
```
No need to put `classical_md` here if your dependencies are already included in the `strict` set of requirements (I should look at copying the abinit dependencies into strict too).

orionarcher replied:
Removed

utf commented on pyproject.toml:
```diff
@@ -59,6 +59,11 @@ forcefields = [
     "torch<=2.2.1",       # incompatibility with dgl if newer versions are used
 ]
+classical_md = [
+    "mdanalysis>=2.7.0",
```
Do these workflows actually require the latest version of mdanalysis etc? If so you can remove these pins from the `classical_md` section. You should keep them for the `strict` section however.

orionarcher replied:
I believe so. They actually require `2.8.0` for some features to work but that hasn't been released yet.

> If so you can remove these pins

A little confused by this, shouldn't I keep the pins if the dependency is required? Did you mean "If not" or am I missing something?

utf commented on pyproject.toml:
```diff
@@ -108,6 +113,9 @@ strict = [
     "torch==2.2.1",
     "typing-extensions==4.12.2",
+    "mdanalysis==2.7.0",
+    "openmm>=8.1.0",
```
This should be pinned with `==`. Also there is a new version available (`8.1.1`)

Andrew-S-Rosen replied:
What in the world..

janosh replied:
@peastman very curious what was the motivation behind removing old versions from pypyi?

Andrew-S-Rosen replied:
Are we sure there were prior PyPI versions to begin with? :)

peastman replied:
There weren't any.  The old package was an ancient front end for running simulations.  It hadn't been supported in many years.

The new packages are still experimental and shouldn't be relied on.  We're still exploring ways of creating PyPI packages.

utf reviewed (commented):
The failing test is very unusual - I wonder if @Andrew-S-Rosen has any idea's as it's related to cclib.

Have you tested installing the dependencies with `pip` rather than `uv`?

orionarcher commented on src/atomate2/openmm/jobs/base.py:
```diff
@@ -0,0 +1,528 @@
+    def make(
+        self,
+        interchange: Interchange | OpenMMInterchange | bytes,
+        prev_task: OpenMMTaskDocument | None = None,
```
Yeah can do. I somewhat dislike having `prev_dir` because you have no idea what needs to be in the directory. But for the sake of minimizing database calls and consistency with other codes, I do think it makes sense.

Andrew-S-Rosen commented:
> The failing test is very unusual - I wonder if @Andrew-S-Rosen has any ideas as it's related to cclib.
> 
> Have you tested installing the dependencies with `pip` rather than `uv`?

Uh... weird! Both `main` and this PR are using the same version of cclib, which hasn't been updated since March. I'm trying to find the source of the issue but am not successful thus far. [I recognize that the cclib interface isn't really used by anything, so it would be a shame for that to hold up a PR...].

orionarcher commented:
@utf is the best way forward to just disable the test here?

orionarcher commented:
And thank you for looking into it @Andrew-S-Rosen

orionarcher commented:
@utf I believe I've addressed all comments and this is ready to merge.

orionarcher commented:
I've pinned the emmet pre-release to get tests passing. @utf are you good to merge this once the new emmet version is released?

janosh commented on src/atomate2/openff/core.py:
```diff
@@ -0,0 +1,165 @@
+    )
+
+    # currently not needed because ForceField isn't correctly supporting iterables
+
```
extra blank line between code and comment?

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+        elif isinstance(spec, MoleculeSpec):
+            mol_specs.append(copy.deepcopy(spec))
+        else:
+            raise TypeError("mol_specs must be a list of dicts or MoleculeSpec")
```
improves UX to print received type here

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+        A list of merged MoleculeSpecs with unique name and SMILES combinations.
+    """
+    mol_specs = copy.deepcopy(mol_specs)
+    merged_spec_dict: dict[tuple[str, str], MoleculeSpec] = {}
```
could use `merged_spec_dict=defaultdict(int)` to avoid the `if key in merged_spec_dict`

orionarcher replied:
My preference is to leave as is, the values are tuples with no sensible default

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+    """
+    # Check if all solvents have corresponding densities
+    solvent_densities = solvent_densities or {}
+    if set(solvents.keys()) > set(solvent_densities.keys()):
```
```suggestion
    if set(solvents) > set(solvent_densities):
```

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+    solvents: dict[str, float],
+    salts: dict[str, float],
+    solvent_densities: dict = None,
+    solvent_ratio_dimension: str = "mass",
```
```suggestion
    solvent_ratio_dimension: Literal["mass", "volume"] = "mass",
```

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+    mol_ratio /= sum(mol_ratio)
+    return {
+        smile: int(np.round(ratio * n_mol))
+        for smile, ratio in zip(species.keys(), mol_ratio)
```
```suggestion
        for smile, ratio in zip(species, mol_ratio)
```

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+    # Convert moles to number of molecules
+    return {
+        smile: int(np.round(ratio * n_mol))
+        for smile, ratio in zip(species.keys(), mol_ratio)
```
```suggestion
        for smile, ratio in zip(species, mol_ratio)
```

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+    total_mass = volume * density  # grams
+
+    # Calculate molecular weights
+    mws = []
```
would use a longer more descriptive name like `mol_weights: list[float] = []`

janosh commented on src/atomate2/openff/utils.py:
```diff
@@ -0,0 +1,304 @@
+            "count": count,
+            "name": name_lookup.get(smile, smile),
+        }
+        if bool(re.search(r"[+-]", smile)):
```
```suggestion
        if re.search(r"[+-]", smile):
```

janosh commented on src/atomate2/openmm/jobs/core.py:
```diff
@@ -0,0 +1,221 @@
+    """
+
+    name: str = "npt simulation"
+    n_steps: int = 1000000
```
```suggestion
    n_steps: int = 1_000_000
```

janosh commented on src/atomate2/openmm/jobs/core.py:
```diff
@@ -0,0 +1,221 @@
+    name : str
+        The name of the NPT simulation job. Default is "npt simulation".
+    n_steps : int
+        The number of simulation steps. Default is 1000000.
```
```suggestion
        The number of simulation steps. Default is 1,000,000.
```

janosh commented on src/atomate2/openmm/jobs/core.py:
```diff
@@ -0,0 +1,221 @@
+        Default is 1 atm.
+    pressure_update_frequency : int
+        The number of steps between pressure
+        update attempts.
```
accidental line break?

janosh commented on src/atomate2/openmm/jobs/core.py:
```diff
@@ -0,0 +1,221 @@
+    """
+
+    name: str = "temperature change"
+    n_steps: int = 1000000
```
same here, more readable with thousands seps imo

janosh commented on src/atomate2/openmm/jobs/generate.py:
```diff
@@ -0,0 +1,323 @@
+            If a molecule is provided, it must have partial charges assigned.
+            If a string is provided, openff_toolkit.Molecule.assign_partial_charges
+            will be used to generate the partial charges.
+
```
blank line

janosh commented on tests/openff_md/conftest.py:
```diff
@@ -0,0 +1,102 @@
+@pytest.fixture
+def temp_dir():
+    with tempfile.TemporaryDirectory() as temp_dir:
+        yield Path(temp_dir)
```
redundant fixture, [`pytest` ships its own `tmp_path`](https://docs.pytest.org/en/stable/how-to/tmp_path.html#the-tmp-path-fixture)

janosh reviewed (changes requested):
huge amount of work here! 👍

orionarcher commented:
This is now pinned to the most recent version of emmet and ready to merge @utf. 

Your previous suggestion to split out OpenMM and OpenFF was very good. They are now independent and the OpenMM workflows support MLFFs and can return structures to better interoperate with the rest of Atomate2.

janosh commented on .github/workflows/testing.yml:
```diff
@@ -31,51 +31,58 @@ jobs:
-      - uses: actions/setup-python@v5
-        with:
-          python-version: ${{ matrix.python-version }}
+      - name: Set up micromamba
```
@utf i think it could help to have the `micromamba`-dependent MD CI be its own job or even its own `test-md.yml` workflow. that would enable only running the MD tests when any MD source files change and would also uncouple the remaining CI from any install/env issues `micromamba` might encounter

utf replied:
I'd be happy with that solution!

orionarcher replied:
Using micromamba saves us from manually building enumlib so I favor keeping it in the main CI.

utf commented:
This really is fantastic. Thank you very much @orionarcher. @janosh, once you're happy I think we can merge.

rkingsbury commented:
Amazing work, thanks for sticking with this @orionarcher !
